stdlib/src/collections/list.mojo

```diff
@@ -239,6 +237,31 @@ struct List[T: CollectionElement](CollectionElement, Sized):
                 self._realloc(self.capacity // 2)
         return ret_val^
 
+    @always_inline
+    fn pop(inout self, i: Int = -1) -> T:
+        """Pops a value from the list at the given index.
+
+        Args:
+            i: The index of the value to pop.
+
+        Returns:
+            The popped value.
+        """
+        debug_assert(-self.size <= i < self.size, "pop index out of range")
```

Comment:
**Thought** This is fine for now.  However, I think we should consider making this API return a `Optional[T]` and embrace algebraic datatypes even though it notably has us differ from Python.  In this particular case, we're _already_ differing from Python since in Python, this function would raise for an index error, whereas this would abort.  This would get changed once we have parametric raises though.  Right now, it's good to _not_ have this function raise until we have parametric raises. 

This is related to [this comment](https://github.com/modularml/mojo/pull/2041#issuecomment-2026512721) from @lsh.

WDYT @modularml/mojo-standard-library?